### PR TITLE
API-4433: Added gateway Connector so notifications will push if there…

### PR DESCRIPTION
**Pull request by mattclark-zerogravit**
… is a push subscriber attached to a box

### CODE REVIEW

File: app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import scala.util.control.NonFatal
+
+@Singleton
+class PushConnector @Inject()(http: HttpClient,
+                              appConfig: AppConfig)
+                             (implicit ec: ExecutionContext) {
+
+  private val outboundHeaders = Seq(
+    (ACCEPT, MimeTypes.JSON),
+    (CONTENT_TYPE, MimeTypes.JSON))
```
Comment (lburgos) [inline app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala:41]: Do we need this? I've never seen it in other connectors and sending json always works. I think it's probably added by http verbs.

File: app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    val msg = "Calling outbound notification gateway"
+    Logger.debug(s"$msg url=${notification.destinationUrl} \nheaders=${hc.headers} \npayload= ${notification.payload}")
+
+    http.POST[OutboundNotification, HttpResponse](url, notification, hc.headers)
+      .map[PushConnectorResult](_ => PushConnectorSuccessResult())
+      .recoverWith {
+        case NonFatal(e) =>
+          e match {
+            case e: BadRequestException => Future.successful(PushConnectorFailedBadRequest(e.message))
+            case _ => Future.successful(PushConnectorFailedResult(e))
```
Comment (lburgos) [inline app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala:60]: Why are we doing this? I can't see the advantage over returning the default response from http verbs (failed future). And it forces you to do pattern matching where you want to use the response value from the connector, whereas with failed futures you can use the response value where you need it (say, in your service), and leave the recovery for wherever you want to specify how to recover from errors (for example, in the controller where you can map errors to HTTP codes).

Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala:60]: it removes exceptions from bubbling up the stack and is more functional. Its probably not obvious without knowing what the gateway responds with but at present the gateway doesn't do much more than validate the payload and so the rest of the errors can be all handled as a failure. If we need the detail later its still there, I doubt we will though. Also the push service is almost fire and forget at present so the connector response is not driving logic. The connector call(s) can fail and the saved notification success response still comes back.

File: app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala (hunk 1 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala:41]: I'll remove, we'll soon find out when we attempt to connect it to the gateway. :)

File: app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala (hunk 2 above)
Comment (mattclark-zerogravit) [reply app/uk/gov/hmrc/pushpullnotificationsapi/connectors/PushConnector.scala:60]: I'll remove the badrequest fail and bundle them all into one.

**Pull request merged**